From 0ccd73ecd7db9caef3e3f5edc12c87383acd0707 Mon Sep 17 00:00:00 2001
From: Tejas Upadhyay <tejas.upadhyay@xilinx.com>
Date: Wed, 27 Jun 2018 02:52:07 -0700
Subject: [PATCH 0380/1852] drm: xlnx: zynqmp_dpsub handling for DP PM

commit 8b02bf388f89fc591d6f16041350cfa2687ccd35 from
https://github.com/Xilinx/linux-xlnx.git

In order to make DisplayPort work after pm suspend/resume,
PHY should exit(suspend) and init(resume) accordingly.

Reviewed-by: Hyun Kwon <hyun.kwon@xilinx.com>
Signed-off-by: Tejas Upadhyay <tejas.upadhyay@xilinx.com>
Signed-off-by: Michal Simek <michal.simek@xilinx.com>
State: pending
Signed-off-by: Yaliang Wang <Yaliang.Wang@windriver.com>
---
 drivers/gpu/drm/xlnx/zynqmp_dp.c    | 24 ++++++++++++++++++++++++
 drivers/gpu/drm/xlnx/zynqmp_dp.h    |  3 ++-
 drivers/gpu/drm/xlnx/zynqmp_dpsub.c | 28 ++++++++++++++++++++++++++++
 3 files changed, 54 insertions(+), 1 deletion(-)

diff --git a/drivers/gpu/drm/xlnx/zynqmp_dp.c b/drivers/gpu/drm/xlnx/zynqmp_dp.c
index d61fb614c136..13dee065d4cb 100644
--- a/drivers/gpu/drm/xlnx/zynqmp_dp.c
+++ b/drivers/gpu/drm/xlnx/zynqmp_dp.c
@@ -470,6 +470,7 @@ void zynqmp_dp_disable_vblank(struct zynqmp_dp *dp)
  *
  * Return: 0 if the phy instances are initialized correctly, or the error code
  * returned from the callee functions.
+ * Note: We can call this function without any phy lane assigned to DP.
  */
 static int zynqmp_dp_init_phy(struct zynqmp_dp *dp)
 {
@@ -542,6 +543,29 @@ static int zynqmp_dp_phy_ready(struct zynqmp_dp *dp)
 	return 0;
 }
 
+/*
+ * Power Management functions
+ */
+/**
+ * zynqmp_dp_pm_resume - Resume DP IP
+ * @dp: DisplayPort IP core structure
+ *
+ * Resume the DP IP including PHY and pipeline.
+ */
+void zynqmp_dp_pm_resume(struct zynqmp_dp *dp)
+{
+	zynqmp_dp_init_phy(dp);
+}
+/**
+ * zynqmp_dp_pm_suspend - Suspend DP IP
+ * @dp: DisplayPort IP core structure
+ *
+ * Suspend the DP IP including PHY and pipeline.
+ */
+void zynqmp_dp_pm_suspend(struct zynqmp_dp *dp)
+{
+	zynqmp_dp_exit_phy(dp);
+}
 /*
  * DP functions
  */
diff --git a/drivers/gpu/drm/xlnx/zynqmp_dp.h b/drivers/gpu/drm/xlnx/zynqmp_dp.h
index 2d7cb8a82fd0..2f6ce3f3e8cf 100644
--- a/drivers/gpu/drm/xlnx/zynqmp_dp.h
+++ b/drivers/gpu/drm/xlnx/zynqmp_dp.h
@@ -27,7 +27,8 @@ void zynqmp_dp_enable_vblank(struct zynqmp_dp *dp);
 void zynqmp_dp_disable_vblank(struct zynqmp_dp *dp);
 void zynqmp_dp_encoder_mode_set_stream(struct zynqmp_dp *dp,
 				       struct drm_display_mode *mode);
-
+void __maybe_unused zynqmp_dp_pm_suspend(struct zynqmp_dp *dp);
+void __maybe_unused zynqmp_dp_pm_resume(struct zynqmp_dp *dp);
 int zynqmp_dp_bind(struct device *dev, struct device *master, void *data);
 void zynqmp_dp_unbind(struct device *dev, struct device *master, void *data);
 
diff --git a/drivers/gpu/drm/xlnx/zynqmp_dpsub.c b/drivers/gpu/drm/xlnx/zynqmp_dpsub.c
index c1a84827a781..8ff251dd5359 100644
--- a/drivers/gpu/drm/xlnx/zynqmp_dpsub.c
+++ b/drivers/gpu/drm/xlnx/zynqmp_dpsub.c
@@ -138,6 +138,33 @@ static int zynqmp_dpsub_remove(struct platform_device *pdev)
 	return err;
 }
 
+static int __maybe_unused zynqmp_dpsub_pm_suspend(struct device *dev)
+{
+	struct platform_device *pdev =
+		container_of(dev, struct platform_device, dev);
+	struct zynqmp_dpsub *dpsub = platform_get_drvdata(pdev);
+
+	zynqmp_dp_pm_suspend(dpsub->dp);
+
+	return 0;
+}
+
+static int __maybe_unused zynqmp_dpsub_pm_resume(struct device *dev)
+{
+	struct platform_device *pdev =
+		container_of(dev, struct platform_device, dev);
+	struct zynqmp_dpsub *dpsub = platform_get_drvdata(pdev);
+
+	zynqmp_dp_pm_resume(dpsub->dp);
+
+	return 0;
+}
+
+static const struct dev_pm_ops zynqmp_dpsub_pm_ops = {
+	SET_SYSTEM_SLEEP_PM_OPS(zynqmp_dpsub_pm_suspend,
+			zynqmp_dpsub_pm_resume)
+};
+
 static const struct of_device_id zynqmp_dpsub_of_match[] = {
 	{ .compatible = "xlnx,zynqmp-dpsub-1.7", },
 	{ /* end of table */ },
@@ -150,6 +177,7 @@ static struct platform_driver zynqmp_dpsub_driver = {
 	.driver			= {
 		.name		= "zynqmp-display",
 		.of_match_table	= zynqmp_dpsub_of_match,
+		.pm             = &zynqmp_dpsub_pm_ops,
 	},
 };
 
-- 
2.31.1

